File: postprocess/nms_on_host.py
"""
Non-Maximum Suppression (NMS) Post-Processing Module for Object Detection.

This module provides functionality for post-processing object detection model outputs
on the host device. It handles the visualization of detection results by drawing
bounding boxes, labels, and confidence scores on images. The module converts
normalized model output coordinates to image coordinates and applies appropriate
visual styling for different object classes.
"""

import colorsys
import json
from typing import Dict, List, Tuple, Union

import cv2
import numpy as np


class ImagePostprocessorNmsOnHost:
    """
    A class for post-processing object detection results and visualizing them on images.

    This class handles the post-processing of object detection model outputs,
    including drawing bounding boxes, labels, and confidence scores on images.
    It converts normalized model output coordinates to image coordinates and
    applies appropriate visual styling for different object classes.

    Attributes:
        pads (Tuple[float, float]): Padding values for x and y coordinates.
        scales (Tuple[float, float]): Scaling factors for x and y coordinates.
        labels (Dict[str, str]): Mapping from class IDs to human-readable labels.
        palette_line (List[Tuple[int, int, int]]): RGB color palette for bounding boxes.
        palette_text (List[Tuple[int, int, int]]): RGB color palette for text labels.
    """

    def __init__(
        self, params: Tuple[Tuple[float, float], Tuple[float, float]], configs: str
    ) -> None:
        """
        Initialize the ImagePostprocessorNmsOnHost with scaling parameters and model configuration.

        Args:
            params (Tuple[Tuple[float, float], Tuple[float, float]]): A tuple containing:
                - scale (Tuple[float, float]): Scaling factors for x and y coordinates.
                - pads (Tuple[float, float]): Padding values for x and y coordinates.
            configs (str): Path to the model configuration JSON file.
        """

        def generate_palette_and_complement(
            num_classes: int,
        ) -> Tuple[List[Tuple[int, int, int]], List[Tuple[int, int, int]]]:
            """
            Generate color palettes for visualization of different object classes.

            Creates two color palettes:
            1. A primary palette for drawing bounding boxes
            2. A complementary palette for drawing text with good contrast

            Args:
                num_classes (int): Number of object classes to generate colors for.

            Returns:
                Tuple[List[Tuple[int, int, int]], List[Tuple[int, int, int]]]:
                    A tuple containing:
                    - rgb_palette: List of RGB colors for drawing bounding boxes
                    - complement_palette: List of complementary RGB colors for text
            """
            # Generate evenly distributed HSV colors for distinct visualization
            # hsv_colors is a List[Tuple[float, float, float]]
            hsv_colors: List[Tuple[float, float, float]] = [
                (i / num_classes, 1.0, 1.0) for i in range(num_classes)
            ]
            # Convert HSV colors to RGB for OpenCV compatibility
            # rgb_palette is a List[Tuple[int, int, int]]
            rgb_palette: List[Tuple[int, int, int]] = [
                tuple(int(c * 255) for c in colorsys.hsv_to_rgb(*hsv))  # c is a float
                for hsv in hsv_colors  # hsv is a Tuple[float, float, float]
            ]

            # Generate complementary colors for better text visibility
            complement_palette: List[Tuple[int, int, int]] = []
            for r, g, b in rgb_palette:
                max_val: int = max(r, g, b)
                min_val: int = min(r, g, b)
                # Calculate complementary color using the max+min-value formula
                comp: Tuple[int, int, int] = (
                    max_val + min_val - r,
                    max_val + min_val - g,
                    max_val + min_val - b,
                )
                complement_palette.append(comp)

            return rgb_palette, complement_palette

        # Load model configuration from JSON file
        with open(configs, "r") as f:  # f is a TextIO
            model_info: List = json.load(f)

        # Extract model configuration and input shape
        model_configs: Dict = model_info[0]
        input_shape: List[int] = model_configs["preprocessing"]["input_shape"][:2]

        # Store padding values for coordinate denormalization
        self.pads: Tuple[float, float] = params[1]
        # Extract scaling factors
        scale: Tuple[float, float] = params[0]
        # Calculate final scales by multiplying input shape with scale factors
        self.scales: Tuple[float, float] = (
            input_shape[0] * scale[0],
            input_shape[1] * scale[1],
        )
        # Get class labels from model info
        self.labels: Dict[str, str] = model_info[1]

        # Generate color palettes for visualization
        self.palette_line: List[Tuple[int, int, int]]
        self.palette_text: List[Tuple[int, int, int]]
        self.palette_line, self.palette_text = generate_palette_and_complement(
            len(self.labels)
        )

    def draw_detections(
        self, frame: np.ndarray, detections: Dict[str, List[np.ndarray]]
    ) -> np.ndarray:
        """
        Draw detection bounding boxes, labels, and confidence scores on the frame.

        Args:
            frame (np.ndarray): The input image frame to draw on.
            detections (Dict[str, List[np.ndarray]]): Detection results from the model.

        Returns:
            np.ndarray: The frame with detection visualizations drawn on it.
        """

        def denormalize_detections(
            detections: Dict[str, List[np.ndarray]],
        ) -> List[List[Union[int, float]]]:
            """
            Convert normalized detection coordinates to image coordinates.

            Args:
                detections (Dict[str, List[np.ndarray]]): Dictionary of detection results.

            Returns:
                List[List[Union[int, float]]]: List of bounding boxes with denormalized coordinates.
                Each box contains [class_id, y_min, x_min, y_max, x_max, confidence].
            """
            # Create scaling factor array for efficient vectorized computation
            array_factor: np.ndarray = np.array(
                [*self.scales, *self.scales], dtype=np.float32
            )
            # Create offset array for padding compensation
            array_offset: np.ndarray = np.array(
                [*self.pads, *self.pads], dtype=np.float32
            )

            # Extract outputs from the detections dictionary
            _, outputs = list(detections.items())[0]

            bounding_boxes: List[List[Union[int, float]]] = []

            # Process each class's detections
            for class_id, output in enumerate(outputs):
                for decoded_bbox in output:
                    # Convert normalized coordinates to image coordinates:
                    # 1. Multiply by scaling factor to get to original image size
                    # 2. Subtract padding to compensate for letterboxing
                    bounding_box: List[Union[int, float]] = [
                        class_id,
                        *(decoded_bbox[:-1] * array_factor - array_offset),
                        decoded_bbox[-1],
                    ]
                    bounding_boxes.append(bounding_box)

            return bounding_boxes

        # Convert normalized coordinates to image coordinates
        detections = denormalize_detections(detections)

        # Draw each detection on the frame
        for detection in detections:
            # Extract class ID from detection
            label_id: int = int(detection[0])

            # Extract confidence score
            confidence: float = float(detection[5])

            # Extract bounding box coordinates
            top_left: Tuple[int, int] = (int(detection[2]), int(detection[1]))
            bottom_right: Tuple[int, int] = (int(detection[4]), int(detection[3]))

            # Draw bounding box rectangle
            cv2.rectangle(
                frame,
                top_left,
                bottom_right,
                color=self.palette_line[label_id],
                thickness=4,
            )

            # Prepare label text with class name and confidence
            label: str = f"{self.labels[str(label_id)]} {confidence:.2f}"
            font_scale: float = 0.8
            text_thickness: int = 2

            # Calculate text size for background rectangle
            (text_width, text_height), baseline = cv2.getTextSize(
                label, cv2.FONT_HERSHEY_SIMPLEX, font_scale, text_thickness
            )

            # Create coordinates for text background rectangle
            bg_top_left: Tuple[int, int] = (top_left[0], top_left[1])
            bg_bottom_right: Tuple[int, int] = (
                top_left[0] + text_width,
                top_left[1] + text_height + baseline,
            )

            # Draw background rectangle for better text visibility
            cv2.rectangle(
                frame,
                bg_top_left,
                bg_bottom_right,
                self.palette_line[label_id],
                -1,  # Filled rectangle
            )

            # Calculate text position and draw the label
            text_org: Tuple[int, int] = (top_left[0], top_left[1] + text_height)
            cv2.putText(
                frame,
                label,
                text_org,
                cv2.FONT_HERSHEY_SIMPLEX,
                font_scale,
                self.palette_text[
                    label_id
                ],  # Use complementary color for better visibility
                text_thickness,
            )

        return frame

    def postprocess(
        self, frame: np.ndarray, outputs: Dict[str, np.ndarray]
    ) -> np.ndarray:
        """
        Post-process model outputs and visualize detections on the frame.

        This is the main entry point for post-processing object detection results.
        It applies visualization if detections are present in the outputs.

        Args:
            frame (np.ndarray): The input image frame to process.
            outputs (Dict[str, np.ndarray]): Model output tensors.

        Returns:
            np.ndarray: The processed frame with detections drawn on it.
        """
        # Only draw detections if outputs are not empty
        if len(outputs) != 0:
            frame = self.draw_detections(frame, outputs)

        return frame

File: cpp/src/postprocess/nms_on_host.cpp
#include "postprocess/nms_on_host.hpp"

#include <fstream>
#include <stdexcept>
#include <cmath>
#include <ranges>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// ---------------------------------------------------------------------------
// Color palette generation
// ---------------------------------------------------------------------------

static cv::Scalar hsv_to_bgr(float h, float s, float v) {
    // h in [0,1], s in [0,1], v in [0,1]
    float r, g, b;
    int hi = static_cast<int>(h * 6.0f) % 6;
    float f = h * 6.0f - std::floor(h * 6.0f);
    float p = v * (1.0f - s);
    float q = v * (1.0f - f * s);
    float t = v * (1.0f - (1.0f - f) * s);

    switch (hi) {
        case 0: r = v; g = t; b = p; break;
        case 1: r = q; g = v; b = p; break;
        case 2: r = p; g = v; b = t; break;
        case 3: r = p; g = q; b = v; break;
        case 4: r = t; g = p; b = v; break;
        default: r = v; g = p; b = q; break;
    }
    return cv::Scalar(b * 255.0f, g * 255.0f, r * 255.0f); // BGR
}

static std::pair<std::vector<cv::Scalar>, std::vector<cv::Scalar>>
generate_palettes(int num_classes) {
    std::vector<cv::Scalar> line_palette, text_palette;
    for (int i = 0; i < num_classes; ++i) {
        float h = static_cast<float>(i) / num_classes;
        cv::Scalar rgb = hsv_to_bgr(h, 1.0f, 1.0f);

        // Complementary: max + min - each channel
        double b = rgb[0], g = rgb[1], r = rgb[2];
        double mx = std::ranges::max({b, g, r});
        double mn = std::ranges::min({b, g, r});
        cv::Scalar comp(mx + mn - b, mx + mn - g, mx + mn - r);

        line_palette.push_back(rgb);
        text_palette.push_back(comp);
    }
    return {line_palette, text_palette};
}

// ---------------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------------

ImagePostprocessorNmsOnHost::ImagePostprocessorNmsOnHost(
    std::pair<std::pair<float,float>, std::pair<int,int>> params,
    const std::string& configs)
{
    std::ifstream f(configs);
    if (!f.is_open())
        throw std::runtime_error("Config file not found: " + configs);

    json model_info;
    try { f >> model_info; }
    catch (const json::parse_error& e) {
        throw std::runtime_error("Error parsing config JSON: " + std::string(e.what()));
    }

    const auto& model_cfg    = model_info[0];
    const auto& input_shape  = model_cfg["preprocessing"]["input_shape"];
    float in_h = input_shape[0].get<float>();
    float in_w = input_shape[1].get<float>();

    pads_   = {static_cast<float>(params.second.first),
               static_cast<float>(params.second.second)};
    scales_ = {in_h * params.first.first,
               in_w * params.first.second};

    const auto& label_map = model_info[1];
    for (auto& [key, val] : label_map.items())
        labels_[key] = val.get<std::string>();

    auto [lp, tp] = generate_palettes(static_cast<int>(labels_.size()));
    palette_line_ = std::move(lp);
    palette_text_ = std::move(tp);
}

// ---------------------------------------------------------------------------
// Denormalization
// ---------------------------------------------------------------------------

std::array<float, 4> ImagePostprocessorNmsOnHost::denormalize_box(
    const std::array<float, 4>& box) const
{
    // box: [y_min, x_min, y_max, x_max]  (normalized 0-1 from model output)
    // scales_: (scale_h, scale_w) = input_h * scale, input_w * scale
    // pads_: (pad_h, pad_w) in original image pixels
    return {
        box[0] * scales_.first  - pads_.first,
        box[1] * scales_.second - pads_.second,
        box[2] * scales_.first  - pads_.first,
        box[3] * scales_.second - pads_.second
    };
}

// ---------------------------------------------------------------------------
// draw_detections
// ---------------------------------------------------------------------------

cv::Mat ImagePostprocessorNmsOnHost::draw_detections(
    cv::Mat frame, const OutputMap& outputs)
{
    if (outputs.empty()) return frame;

    const auto& [first_name, first_output] = *outputs.begin();
    if (!first_output.is_nms) return frame;

    const auto& nms = first_output.nms;

    for (size_t class_id = 0; class_id < nms.size(); ++class_id) {
        for (const auto& det : nms[class_id]) {
            // Denormalize
            std::array<float, 4> box = {det.y_min, det.x_min, det.y_max, det.x_max};
            auto db = denormalize_box(box);

            cv::Point top_left(static_cast<int>(db[1]), static_cast<int>(db[0]));
            cv::Point bottom_right(static_cast<int>(db[3]), static_cast<int>(db[2]));

            const cv::Scalar& line_color = palette_line_[class_id % palette_line_.size()];
            const cv::Scalar& text_color = palette_text_[class_id % palette_text_.size()];

            // Draw bounding box
            cv::rectangle(frame, top_left, bottom_right, line_color, 4);

            // Prepare label
            auto it = labels_.find(std::to_string(class_id));
            std::string class_name = (it != labels_.end()) ? it->second : "unknown";
            char conf_buf[32];
            std::snprintf(conf_buf, sizeof(conf_buf), "%.2f", det.confidence);
            std::string label = class_name + " " + conf_buf;

            const double font_scale = 0.8;
            const int text_thickness = 2;
            int baseline = 0;
            cv::Size text_sz = cv::getTextSize(
                label, cv::FONT_HERSHEY_SIMPLEX, font_scale, text_thickness, &baseline);

            cv::Point bg_tl = top_left;
            cv::Point bg_br(top_left.x + text_sz.width,
                             top_left.y + text_sz.height + baseline);
            cv::rectangle(frame, bg_tl, bg_br, line_color, cv::FILLED);

            cv::putText(frame, label,
                        cv::Point(top_left.x, top_left.y + text_sz.height),
                        cv::FONT_HERSHEY_SIMPLEX, font_scale,
                        text_color, text_thickness);
        }
    }
    return frame;
}

// ---------------------------------------------------------------------------
// postprocess (entry point)
// ---------------------------------------------------------------------------

cv::Mat ImagePostprocessorNmsOnHost::postprocess(
    const cv::Mat& frame, const OutputMap& outputs)
{
    if (outputs.empty()) return frame;
    return draw_detections(frame.clone(), outputs);
}
